### ISSUE TITLE
illegal cyclic reference involving type A

### ISSUE THREAD

**Issue by ponythewhite** [scalamacros/paradise#42]
The following line in sbt build:

``` scala
addCompilerPlugin("org.scalamacros" % "paradise" % "2.0.0-SNAPSHOT" cross CrossVersion.full)
```

causes the following code:

``` scala
  import newtype
  import newtype._

  type Timestamp[@specialized(Int, Long) A] = Newtype[A, TimestampOps[A]]
  def Timestamp[@specialized(Int, Long) A](i: A): Timestamp[A] = newtype(i)
  implicit class TimestampOps[@specialized(Int, Long) A](val i: A){
}

object newtype {
  import scala.language.implicitConversions
  def apply[Repr, Ops](r : Repr) : Newtype[Repr, Ops] = r.asInstanceOf[Any with Newtype[Repr, Ops]]
  type Newtype[Repr, Ops] = { type Tag = NewtypeTag[Repr, Ops] }
  trait NewtypeTag[Repr, Ops]
  implicit def newtypeOps[Repr, Ops](t : Newtype[Repr, Ops])(implicit mkOps : Repr => Ops) : Ops = t.asInstanceOf[Repr]
}
```

to throw:

```
[error]  illegal cyclic reference involving type A
[error]   implicit class TimestampOps[@specialized(Int, Long) A](val i: A) {
[error]                                                       ^
```

Checked with Scala 2.10.3 & 2.10.4 Without `@specialized(Int, Long)` the code compiles.

As a workaround I've moved the `addCompilerPlugin` to the macro subproject.

**Comment by xeno-by** [comment]
Minimized repro:

```
object M {
  implicit class TimestampOps[@specialized(Int, Long) A](val i: A){}
}
```

Errors out both on 2.10.x and master.

**Comment by xeno-by** [comment]
@ponythewhite I think I have identified and fixed the problem, but it's in such a core area of the plugin that I'd prefer to delay the fix until after 2.0.0-final is released. Would it be okay for you?

**Comment by ponythewhite** [comment]
@xeno-by a temporary workaround of splitting the project into distinct sbt configurations with and without the sbt plugin works for now. This is not critical for me, so the fix being included in 2.0.0-final will be fine.

Many thanks for prompt feedback!

**Comment by xeno-by** [comment]
@ponythewhite Sorry for the confusion. What I meant is that I will probably have to release 2.0.0-final with this as a known issue in order to avoid last-minute breakages. Along with that, I'm also planning to release 2.1.0-M1 with this fix included. Hope this is going to be okay.

**Comment by ponythewhite** [comment]
@xeno-by ah yes, still works for me! Thanks.

**Comment by xeno-by** [comment]
Fixed in 2.1.0-SNAPSHOT, which will be soon deployed to sonatype by our jenkins.
